web/page0/src/app/device-form.service.ts

```diff
@@ -70,20 +86,20 @@ export class DeviceFormService {
     });
   }
 
-  private toFormArray(deviceSettings: DeviceSetting[]) {
+  private toFormArray(deviceSettings: DeviceSetting[]): DeviceForm {
     return this.formBuilder.array(
       deviceSettings.map(setting => this.toDeviceForm(setting))
     );
   }
 
-  private getInitDeviceForm() {
+  private getInitDeviceForm(): DeviceForm {
     return this.toFormArray(this.getInitDeviceSettings());
   }
 
   private deviceSettingsForm$ = this.deviceFormAction$.pipe(
-    tap(action => console.log('deviceForm ', action.type)),
+    tap((action: DeviceFormAction) => console.log('deviceForm ', action.type)),
```

Comment:
It might be better to remove console.logs for debugging after development is completed.